core/lib/zksync_core/src/basic_witness_input_producer/mod.rs

```diff
@@ -0,0 +1,185 @@
+use std::sync::Arc;
+use std::time::Instant;
+
+use zksync_dal::ConnectionPool;
+use zksync_object_store::{ObjectStore, ObjectStoreFactory};
+use zksync_queued_job_processor::JobProcessor;
+use zksync_types::witness_block_state::WitnessBlockState;
+use zksync_types::{L1BatchNumber, L2ChainId};
+
+use async_trait::async_trait;
+use tokio::runtime::Handle;
+use tokio::task::JoinHandle;
+
+mod vm_interactions;
+
+use vm_interactions::{create_vm, execute_tx};
+use zksync_types::block::MiniblockExecutionMode;
+
+/// Component that extracts all data (from DB) necessary to run a Basic Witness Generator.
+/// This component will upload Witness Inputs to the object store.
+/// This allows Witness Generator workflow (that needs only Basic Witness Generator Inputs)
+/// to be run only using the object store information, having no other external dependency.
+#[derive(Debug)]
+pub struct BasicWitnessInputProducer {
+    connection_pool: ConnectionPool,
+    validation_computational_gas_limit: u32,
+    l2_chain_id: L2ChainId,
+    object_store: Arc<dyn ObjectStore>,
+}
+
+impl BasicWitnessInputProducer {
+    pub async fn new(
+        connection_pool: ConnectionPool,
+        store_factory: &ObjectStoreFactory,
+        l2_chain_id: L2ChainId,
+    ) -> anyhow::Result<Self> {
+        let validation_computational_gas_limit = u32::MAX;
+        Ok(BasicWitnessInputProducer {
+            connection_pool,
+            validation_computational_gas_limit,
+            object_store: store_factory.create_store().await.into(),
+            l2_chain_id,
+        })
+    }
+
+    fn process_job_impl(
+        rt_handle: Handle,
+        l1_batch_number: L1BatchNumber,
+        started_at: Instant,
+        connection_pool: ConnectionPool,
+        validation_computational_gas_limit: u32,
+        l2_chain_id: L2ChainId,
+    ) -> anyhow::Result<WitnessBlockState> {
+        let mut connection = rt_handle.block_on(connection_pool.access_storage())?;
+
+        let miniblocks_execution_data = rt_handle.block_on(
+            connection
+                .transactions_dal()
+                .get_miniblocks_to_execute_for(MiniblockExecutionMode::L1Batch(l1_batch_number)),
+        );
+
+        let (mut vm, storage_view) = create_vm(
+            rt_handle.clone(),
+            l1_batch_number,
+            connection,
+            validation_computational_gas_limit,
+            l2_chain_id,
+        );
+
+        tracing::info!("Started execution of l1_batch: {l1_batch_number:?}");
+        for (index, miniblock_execution_data) in miniblocks_execution_data.iter().enumerate() {
+            tracing::debug!(
+                "Started execution of miniblock: {:?}",
+                miniblock_execution_data.number
+            );
+            for tx in &miniblock_execution_data.txs {
+                tracing::debug!("Started execution of tx: {tx:?}");
```

Comment:
Nit: This could spam logs (of course, could be configured away, but still); I'd suggest to use the `TRACE` level here.